language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToLocalVarCodeAction.java

```diff
@@ -0,0 +1,136 @@
+public class ExtractToLocalVarCodeAction implements RangeBasedCodeActionProvider {
+
+    public static final String NAME = "Extract To Local Variable";
+    private static final String VARIABLE_NAME_PREFIX = "variable";
+
+    public List<SyntaxKind> getSyntaxKinds() {
+        return List.of(SyntaxKind.BOOLEAN_LITERAL, SyntaxKind.NUMERIC_LITERAL, SyntaxKind.STRING_LITERAL,
+                SyntaxKind.BINARY_EXPRESSION, SyntaxKind.START_ACTION, SyntaxKind.BRACED_EXPRESSION,
+                SyntaxKind.FUNCTION_CALL, SyntaxKind.QUALIFIED_NAME_REFERENCE, SyntaxKind.INDEXED_EXPRESSION,
+                SyntaxKind.FIELD_ACCESS, SyntaxKind.METHOD_CALL, SyntaxKind.CHECK_EXPRESSION, 
+                SyntaxKind.MAPPING_CONSTRUCTOR, SyntaxKind.TYPEOF_EXPRESSION, SyntaxKind.UNARY_EXPRESSION,
+                SyntaxKind.TYPE_TEST_EXPRESSION, SyntaxKind.TRAP_EXPRESSION, SyntaxKind.LIST_CONSTRUCTOR, 
+                SyntaxKind.TYPE_CAST_EXPRESSION, SyntaxKind.TABLE_CONSTRUCTOR, SyntaxKind.LET_EXPRESSION,
+                SyntaxKind.IMPLICIT_NEW_EXPRESSION, SyntaxKind.EXPLICIT_NEW_EXPRESSION, 
+                SyntaxKind.PARENTHESIZED_ARG_LIST, SyntaxKind.ERROR_CONSTRUCTOR, SyntaxKind.OBJECT_CONSTRUCTOR);
+    }
+
+    @Override
+    public boolean validate(CodeActionContext context, RangeBasedPositionDetails positionDetails) {
+        return CodeActionNodeValidator.validate(context.nodeAtRange());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<CodeAction> getCodeActions(CodeActionContext context,
+                                           RangeBasedPositionDetails posDetails) {
+        
+        if (context.currentSyntaxTree().isEmpty() || context.currentSemanticModel().isEmpty()) {
+            return Collections.emptyList();
+        }
+        
+        Node node = posDetails.matchedCodeActionNode(); 
+        if ((node.kind() == SyntaxKind.MAPPING_CONSTRUCTOR && node.parent().kind() == SyntaxKind.TABLE_CONSTRUCTOR) 
+                || (node.kind() == SyntaxKind.FUNCTION_CALL && node.parent().kind() == SyntaxKind.LOCAL_VAR_DECL)) {
+            return Collections.emptyList();
+        }
+        
+        
+        String varName = getLocalVarName(context);
+        String value = node.toSourceCode().strip();
+        LineRange replaceRange = node.lineRange();
+        Optional<TypeSymbol> typeSymbol = context.currentSemanticModel().get().typeOf(node);
+        if (typeSymbol.isEmpty() || typeSymbol.get().typeKind() == TypeDescKind.COMPILATION_ERROR) {
+            return Collections.emptyList();
+        }
+
+        Node statementNode = node;
+        while (statementNode != null && !(statementNode instanceof StatementNode) 
+                && !(statementNode instanceof ModuleMemberDeclarationNode)) {
+            statementNode = statementNode.parent();
+        }
+        if (statementNode == null) {
+            return Collections.emptyList();
+        }
```

Comment:
Shall we extract this segment to a method to get the statement node?

Reply:
Done.